src/system/random/basic.lean

```diff
@@ -0,0 +1,479 @@
+import algebra.group_power
+import category.liftable
+
+import data.bitvec
+import data.list.basic
+import data.stream
+import tactic.cache
+import data.fin
+
+import tactic.interactive
+import tactic.norm_num
+
+import system.io
+import system.random
+
+open list io applicative
+
+universes u v w
+
+/-- A monad to generate random objects using the generator type `g` -/
+@[reducible]
+def rand_g (g : Type) := state (ulift g)
+/-- A monad to generate random objects using the generator type `std_gen` -/
+@[reducible]
+def rand := rand_g std_gen
+
+instance (g : Type) : liftable (rand_g.{u} g) (rand_g.{v} g) :=
+@state_t.liftable' _ _ _ _ _ (equiv.ulift.trans.{u u u u u} equiv.ulift.symm)
+
+open ulift (hiding inhabited)
+
+/-- Generate one more `ℕ` -/
+def random.next {gen : Type} [random_gen gen] : rand_g gen ℕ :=
+⟨ prod.map id up ∘ random_gen.next ∘ down ⟩
+
+/-- `range i j` is a number between `i` and `j` inclusively -/
+@[nolint has_inhabited_instance]
+def range {α : Type u} [has_le α] (i j : α) :=
+{ x : α // i ≤ x ∧ x ≤ j }
+
+infix ` .. `:41 := range
+
+open stream
+
+/-- `random α` gives us machinery to generate values of type `α` -/
+class random (α : Type u) extends has_le α :=
+(random : Π (g : Type) [random_gen g], rand_g g α)
+(random_r : Π g [random_gen g] (x y : α),
+              x ≤ y →
+              rand_g g (x .. y))
+(random_series : Π (gen : Type) [random_gen gen], gen → stream α :=
```

Comment:
I'm curious why you wrote these as structure fields with default values, rather than separate declarations. Will someone instantiating these ever override them?